crates/turborepo-lib/src/run/watch.rs

```diff
@@ -76,43 +104,56 @@ impl WatchClient {
             execution_args: execution_args.clone(),
         });
 
-        let mut run = RunBuilder::new(new_base)?
+        let run = RunBuilder::new(new_base)?
             .build(&handler, telemetry.clone())
             .await?;
 
-        run.print_run_prelude();
-
         let connector = DaemonConnector {
             can_start_server: true,
             can_kill_server: true,
             paths: DaemonPaths::from_repo_root(&base.repo_root),
         };
 
+        Ok(Self {
+            base,
+            run,
+            connector,
+            handler,
+            telemetry,
+            persistent_tasks_handle: None,
+        })
+    }
+
+    pub async fn start(&mut self) -> Result<(), Error> {
+        let connector = self.connector.clone();
         let mut client = connector.connect().await?;
 
         let mut events = client.package_changes().await?;
-        let mut current_runs: HashMap<PackageName, JoinHandle<Result<i32, run::Error>>> =
-            HashMap::new();
-        let mut persistent_tasks_handle = None;
+
+        self.run.print_run_prelude();
+
+        let signal_subscriber = self.handler.subscribe().ok_or(Error::NoSignalHandler)?;
+
+        let changed_packages = Mutex::new(RefCell::new(ChangedPackages::default()));
 
         let event_fut = async {
             while let Some(event) = events.next().await {
                 let event = event?;
-                Self::handle_change_event(
-                    &mut run,
-                    event.event.unwrap(),
-                    &mut current_runs,
-                    &base,
-                    &telemetry,
-                    &handler,
-                    &mut persistent_tasks_handle,
-                )
-                .await?;
+                Self::handle_change_event(&changed_packages, event.event.unwrap()).await?;
             }
 
             Err(Error::ConnectionClosed)
         };
 
+        let run_fut = async {
+            loop {
+                if !changed_packages.lock().unwrap().borrow().is_empty() {
+                    self.execute_run(&changed_packages).await?;
+                }
+                yield_now().await;
```

Comment:
Can you document what's going on here? I'm guessing it's to avoid spinning and not receiving any new events...

Maybe there's a better way to structure that? Like a watch or something so we can block until there's a change?

Reply:
Yeah we could probably do this with a watch channel. Will give that a shot

Reply:
I think with the watch, we don't need this line anymore, right?